distributed/deploy/cluster.py

```diff
@@ -564,3 +565,71 @@ def __eq__(self, other):
 
     def __hash__(self):
         return id(self)
+
+    async def _wait_for_workers(self, n_workers=0, timeout=None):
+        await self.update_scheduler_info(
+            comm_name="Cluster worker status", op="identity"
+        )
+        info = SchedulerInfo(self.scheduler_info)
+        self._scheduler_identity = info
+        if timeout:
+            deadline = time() + parse_timedelta(timeout)
+        else:
+            deadline = None
+
+        def running_workers(info):
+            return len(
+                [
+                    ws
+                    for ws in info["workers"].values()
+                    if ws["status"] == Status.running.name
+                ]
+            )
+
+        while n_workers and running_workers(info) < n_workers:
+            if deadline and time() > deadline:
+                raise TimeoutError(
+                    "Only %d/%d workers arrived after %s"
+                    % (running_workers(info), n_workers, timeout)
+                )
+            await asyncio.sleep(0.1)
+
+            await self.update_scheduler_info(
+                comm_name="Cluster worker status", op="identity"
+            )
+            info = SchedulerInfo(self.scheduler_info)
+            self._scheduler_identity = info
+
+    async def update_scheduler_info(self, comm_name: str, op: str) -> Comm:
+        """Send comm to scheduler requesting information and update scheduler_info accordingly
+
+        Parameters
+        ----------
+        comm_name : string
+            Name of communication, e.g. "Cluster worker status"
+        op : string
+            Name of operation, e.g. "identity"
+
+        Returns
+        -------
+        comm : distributed.comm.tcp.TCP
+            Comm to the scheduler
+        """
+        comm = await self.scheduler_comm.live_comm()
+        comm.name = comm_name
+        await comm.write({"op": op})
+        self.scheduler_info = await comm.read()
+        return comm
```

Comment:
This still feels strange, why would a method called `update_scheduler_info` take a `comm` and return a `live_comm` that then gets reused by the `_watch_worker_status`?

I the `_start` we should create the comm and do whatever we need to do with it. Then in `update_scheduler_info` we should just use that existing comm. Maybe part of the issue here is that we don't assign `comm` to anything in `_start` so can't reuse it later.

Reply:
I tried assigning comm to an attribute and reusing it, but it raised an assertion error about reading for a comm that was already being read, I assume as it's being used to wait for workers being added/removed. Instead I created a new comm in _start and use that whenever we need to update scheduler_info. It seems to work fine, and update_scheduler_info is now much simpler. Let me know what you think

Reply:
Thanks for tackling this, it definitely reads a bit more clearly now.

It'd be great to get a review from @fjetter here on the implications of handling multiple comms this way.